src/main/java/org/testng/GuiceHelper.java

```diff
@@ -33,15 +40,38 @@ Injector getInjector(IClass iClass) {
     List<Module> moduleInstances =
         Lists.newArrayList(getModules(guice, parentInjector, iClass.getRealClass()));
 
-    // Reuse the previous injector, if any
+    if (isStringNotEmpty(context.getSuite().getParentModule())) {
+      Class<Module> parentModule = (Class<Module>) ClassHelper.forName(context.getSuite().getParentModule());
+      if (parentModule == null) {
+        throw new TestNGException(
+                "Cannot load parent Guice module class: " + context.getSuite().getParentModule());
+      }
+      moduleInstances.add(newParentModule(parentModule));
+    }
+
+    // Reuse the previous injector, if any, but don't create a child injector as JIT bindings can conflict
     Injector injector = context.getInjector(moduleInstances);
     if (injector == null) {
-      injector = parentInjector.createChildInjector(moduleInstances);
+      Stage stage = Stage.DEVELOPMENT;
+      String stageString = context.getSuite().getGuiceStage();
+      if (isStringNotEmpty(stageString)) {
+        stage = Stage.valueOf(stageString);
```

Comment:
Maybe you can factorize some code from https://github.com/cbeust/testng/blob/55d3ebd42be1d5eba7b6f374245a986c249c4291/src/main/java/org/testng/internal/ClassImpl.java#L135-L154

The most logic will be that ClassImpl will use a method of GuiceHelper

Reply:
Yep, refactoring is better than copy/paste, I was just being overly cautious with scope of the change.